### PR TITLE
Add CheckedPopupMenuitem, update the Stocks demo

### LINKED ISSUE #2200: Add CheckedPopupMenuItem
The gallery MenuDemo creates this by hand. For convenience's sake they should be a simple subclass of PopupMenuItem.

Note that this is not a "checkbox" menu item, see https://www.google.com/design/spec/components/menus.html#menus-menu-items

### PR DESCRIPTION
**Pull request by HansMuller**
Using PopupMenuButton and CheckedPopupMenuItem in the stocks demo.

Fixes https://github.com/flutter/flutter/issues/2200

Also:
- Render disabled menu item icons correctly
- Rename the disabled property to enabled (and etc)
- Add an enabled property to ListItem
- Don't call PopupMenuItem.onSelected unless an item was actually selected

### CODE REVIEW

File: examples/material_gallery/lib/demo/menu_demo.dart [hunk 1]
```diff
@@ -94,7 +94,7 @@ class MenuDemoState extends State<MenuDemo> {
                   child: new Text('Context menu item one')
                 ),
                 new PopupMenuItem(
-                  disabled: true,
+                  isDisabled: true,
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/menu_demo.dart]: Can we change this to "enabled: false"? The Web has had a number of issues due to having booleans that are negative adjectives.

Comment (Hixie) [inline examples/material_gallery/lib/demo/menu_demo.dart]: It's a pity we can't use the onPressed != null approach.

Comment (Hixie) [inline examples/material_gallery/lib/demo/menu_demo.dart]: maybe value != null?

Comment (HansMuller) [inline examples/material_gallery/lib/demo/menu_demo.dart]: > Can we change this to "enabled: false"? The Web has had a number of issues due to having booleans that are negative adjectives.

Sure. I've been using isFoo for boolean properties, so isEnabled?

Comment (HansMuller) [inline examples/material_gallery/lib/demo/menu_demo.dart]: > maybe value != null?

I have mixed feelings about this kind of overloading since developers are unlikely to guess that it's how one disables an item. But if you really think it's a good idea, OK.

Comment (Hixie) [inline examples/material_gallery/lib/demo/menu_demo.dart]: let's stick with the boolean.

As far as isFoo vs foo, for widget constructor arguments, we don't generally do the is prefix. isFoo is usually reserved for getters that examine state.

File: examples/stocks/lib/stock_home.dart [hunk 2]
```diff
@@ -59,24 +92,31 @@ class StockHomeState extends State<StockHome> {
...
-      onAutorefreshChanged: _handleAutorefreshChanged
-    );
+  void _handleStockMenu(BuildContext context, _StockMenuItem value) {
+    switch(value) {
+      case _StockMenuItem.autorefresh:
+        setState(() {
+          _autorefresh = !_autorefresh;
+        });
+        break;
+      case _StockMenuItem.refresh:
```
Comment (Hixie) [inline examples/stocks/lib/stock_home.dart]: does refresh with now?

File: examples/stocks/lib/stock_home.dart [hunk 3]
```diff
@@ -59,24 +92,31 @@ class StockHomeState extends State<StockHome> {
...
+        showDialog(
+          context: context,
+          child: new _NotImplementedDialog()
```
Comment (Hixie) [inline examples/stocks/lib/stock_home.dart]: why not inline the dialog?

Comment (HansMuller) [inline examples/stocks/lib/stock_home.dart]: The dialog seems like a hack, plus it inexplicably doesn't show up. I was anticipating that the next step would be to just remove it. After making it work.

File: examples/stocks/lib/stock_home.dart (hunk 2 above)
Comment (HansMuller) [inline examples/stocks/lib/stock_home.dart]: No. The call to showDialog() happens but, as before, the dialog does not appear.

File: examples/material_gallery/lib/demo/menu_demo.dart (hunk 1 above)
Comment (HansMuller) [inline examples/material_gallery/lib/demo/menu_demo.dart]: OK. I'll de-prefix ListItem's isThreeLine, isDisabled, and isDense as well.

File: examples/stocks/lib/stock_home.dart (hunk 2 above)
Comment (Hixie) [inline examples/stocks/lib/stock_home.dart]: ok, cool. (I want to make sure that when we fix this we understand what was wrong.)

File: examples/stocks/lib/stock_home.dart (hunk 3 above)
Comment (Hixie) [inline examples/stocks/lib/stock_home.dart]: Right, that's why I would ahve inlined it. Giving it a class makes it look less hacky. ;-)

It's fine either way.

File: packages/flutter/lib/src/material/list_item.dart [hunk 4]
```diff
@@ -36,11 +37,17 @@ class ListItem extends StatelessComponent {
   final Widget right;
   final bool isThreeLine;
   final bool isDense;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/list_item.dart]: Just a thought, not a comment on this patch: maybe we should convert isThreeLine and isDense into a single enum that gives the various ways that this can be laid out.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 5]
```diff
@@ -49,18 +69,31 @@ class PopupMenuItem<T> extends StatelessComponent {
...
       )
     );
   }
 }
 
+class CheckedPopupMenuItem<T> extends PopupMenuItem<T> {
+  CheckedPopupMenuItem({
+    Key key,
+    T value,
+    isChecked: false,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/popup_menu.dart]: just "checked" would be more idiomatic, I think.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 6]
```diff
@@ -49,18 +69,31 @@ class PopupMenuItem<T> extends StatelessComponent {
...
+    bool isDisabled: false,
+    Widget child
+  }) : super(
+    key: key,
+    value: value,
+    isDisabled: isDisabled,
+    child: new ListItem(
+      isDisabled: isDisabled,
+      left: new Icon(icon: isChecked ? 'action/done' : null),
```
Comment (Hixie) [inline packages/flutter/lib/src/material/popup_menu.dart]: shouldn't the check mark animate when the item is checked?

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 7]
```diff
@@ -21,27 +25,43 @@ const double _kMenuMinWidth = 2.0 * _kMenuWidthStep;
...
 
 class PopupMenuItem<T> extends StatelessComponent {
   PopupMenuItem({
     Key key,
     this.value,
-    this.disabled: false,
+    this.isDisabled: false,
     this.hasDivider: false,
     this.child
   }) : super(key: key);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/popup_menu.dart]: maybe assert that the value isn't null if the item is enabled?

**Comment by Hixie** [comment]
LGTM. Sorry about not catching the isFoo thing in earlier patches.

File: examples/material_gallery/lib/demo/menu_demo.dart (hunk 1 above)
Comment (Hixie) [inline examples/material_gallery/lib/demo/menu_demo.dart]: "isThreeLine" is probably better than "threeLine" since the latter isn't an adjective. Generally I think my intuition is that arguments that are booleans should be adjective, and arguments that take other types should be nouns. (length:3 -> "the length is 3" vs dens:true -> "this is dense")

File: packages/flutter/lib/src/material/list_item.dart (hunk 4 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/list_item.dart]: It would be nice to just have one layoutStyle parameter. It might end up looking like:

ListItemLayoutStyle {oneLine, twoLine, threeLine, denseOneLine, denseTwoLine, denseThreeLine}

File: packages/flutter/lib/src/material/popup_menu.dart (hunk 6 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/popup_menu.dart]: I would think so although I couldn't find evidence in the MD spec (I've asked about it).

Assuming that the animation is needed, I'd like to make that change in a separate CL, if that's OK.

File: packages/flutter/lib/src/material/popup_menu.dart (hunk 5 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/popup_menu.dart]: Yes, I'm going to remove all of the "is" prefixes I've added in the last 24 hours or so :-).

File: packages/flutter/lib/src/material/popup_menu.dart (hunk 7 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/popup_menu.dart]: That sounds like a good mistake catcher.  Will do.

File: examples/stocks/lib/stock_home.dart (hunk 2 above)
Comment (HansMuller) [inline examples/stocks/lib/stock_home.dart]: Just FTR: https://github.com/flutter/flutter/issues/1099

File: examples/material_gallery/lib/demo/menu_demo.dart (hunk 1 above)
Comment (HansMuller) [inline examples/material_gallery/lib/demo/menu_demo.dart]: I've changed isDense to dense. Sadly, I've just noticed that Input also supports dense layout. It's property is called isDense. Sigh.
